data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/authenticator/DynamicBasicCredentialsProvider.java

```diff
@@ -0,0 +1,51 @@
+package org.opensearch.dataprepper.plugins.kafka.authenticator;
+
+import org.opensearch.dataprepper.plugins.kafka.configuration.AuthConfig;
+import org.opensearch.dataprepper.plugins.kafka.configuration.KafkaConnectionConfig;
+import org.opensearch.dataprepper.plugins.kafka.configuration.PlainTextAuthConfig;
+
+import java.util.Objects;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class DynamicBasicCredentialsProvider {
+    private static final DynamicBasicCredentialsProvider singleton = new DynamicBasicCredentialsProvider();
+
+    public static DynamicBasicCredentialsProvider getInstance() {
+        return singleton;
+    }
+
+    private BasicCredentials basicCredentials;
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+    // Used for testing only
+    protected DynamicBasicCredentialsProvider() {}
+
+    public BasicCredentials getBasicCredentials() {
```

Comment:
Is `DynamicSaslClientCallbackHandler::handle` and thus this method called for every interaction with Kafka? 

I'm wondering what overhead the read/write lock will have here. And if it is necessary. If we make a request and the username/password is stale, is it that big of a deal? The user needs to setup a proper rotation. Or there will more failures while waiting to get the updated secrets than the difference between two requests to read the value.

Reply:
> Is DynamicSaslClientCallbackHandler::handle and thus this method called for every interaction with Kafka?
Yes. 

> If we make a request and the username/password is stale, is it that big of a deal?
w/o readwrite lock, I am not sure how stale the credentials could be. 

In my POC sasl handshake happens in the background decoupled from individual API call.